### PR TITLE
RFC: improve ACK filtering safety

### PR DESCRIPTION
**Pull request by rmounce**
I think this should address Eric's concerns, however more eyeballs appreciated.

Compile & run tested on Ubuntu 18.04 amd64.

### CODE REVIEW

File: sch_cake.c [hunk 1]
```diff
@@ -875,12 +875,14 @@ static struct sk_buff *cake_ack_filter(struct cake_sched_data *q,
 	ipv6h = skb->encapsulation ? inner_ipv6_hdr(skb) : ipv6_hdr(skb);
 
 	/* check that the innermost network header is v4/v6, and contains TCP */
-	if (iph->version == 4) {
+	if (pskb_may_pull(skb, ((unsigned char *)iph - skb->head) + sizeof(struct iphdr)) &&
+	    iph->version == 4) {
```
Comment (tohojo) [inline sch_cake.c:879]: Maybe reverse these so logic short-circuiting dosn't call pskb_may_pull if it's not ipv4? (and likewise below)

Comment (rmounce) [reply sch_cake.c:879]: I’m not sure if there’s a case where there may not be any transport header, which the current order would defend against.

Comment (tohojo) [reply sch_cake.c:879]: Ah right, of course. Silly me. That makes me wonder if the encapsulation check above could be problematic?

**Comment by tohojo** [comment]
BTW, while you're poking at this, it may be worth using the kernel helpers to get to the headers instead of the raw pointer arithmetic currently used to get to the tcp header. See example here:

https://git.kernel.org/pub/scm/linux/kernel/git/davem/net-next.git/tree/net/core/dev.c#n3187 (the use of skb_header_pointer in particular).

**Comment by tohojo** [comment]
Actually, I think skb_header_pointer already does the bounds checking we want...

**Comment by tohojo** [comment]
Ah, but it won't defragment a fragmented packet...

**Comment by rmounce** [comment]
My rationale for using pointer arithmetic to get the TCP header was to try and make ACK filtering as robust as possible with e.g. 6in4 traffic, or cases where netfilter isn't used on a middlebox and (inner_)transport_header isn't set (if this is even possible?)

This may be making things harder than they need to be, but it should now hopefully be memory safe.

**Pull request merged**